### DISCUSSION TITLE
How to conduct validation test during training with multi GPU?

### DISCUSSION THREAD

**Discussion by yjiangling** [horovod/horovod#3095] [Q&A]
Hi all,

        When I use multi GPUs to train, but I want to conduct validation test during the train, how can I realize it? Here is my code:

`
with tf.Session(config=config) as sess:
    ckpt = tf.train.latest_checkpoint(hp.checkpoint)

    if ckpt is None:
        logging.info("Starting new training")
        sess.run(tf.global_variables_initializer())
        sess.run(bcast)
    else:
        logging.info("Resuming from checkpoint: %s" % ckpt)
        saver.restore(sess, ckpt)

    while True:
        try:
            ids, _gs, _loss, _acc, _summary, _ = sess.run([train_id, global_step, loss, accuracy, train_summary, train_op])
            if hvd.rank() == 0:
                logging.info("step {}, loss:{:.4f}, accuracy:{:.4f}".format(_gs, _loss, _acc))

                if _gs % 10 == 0:
                    summary_writer.add_summary(_summary, _gs)

                if _gs % 1000 == 0:
                    logging.info("# save models at {} step".format(_gs))
                    saver.save(sess, ckpt_name, global_step=_gs)

                if math.isnan(_loss):
                    logging.info("第hvd.rank({})个进程的第{}步出现错误".fromat(ids, _gs))
                    raise Exception('Loss Exploded')

                if _gs % hp.eval_per_step == 0:
                    logging.info("# statrt a validation test: ")
                    ......
`

It give the error as this:

INFO:root:Horovod has been shut down. This was caused by an exception on one of the ranks or an attempt to allreduce, allgather or broadcast a tensor after one of the ranks finished execution. If the shutdown was caused by an exception, you should see the exception in the log before the first shutdown message.
[[Node:DistributedAdamOptimizer_Allreduce/HorovodAllreduce_gradients_encoder_dense_Tensordot_transpose_1_grad_transpose_0 = HorovodAllreduce[T=DT_FLOAT, _device="/job:localhost/replica:0/task:0/device:GPU:0"](gradients/encoder/dense/Tensordot/transpose_1_grad/transpose)]]

Anyone can give some suggestion? Thanks a lot!

**Comment by maxhgerlach** [comment]
All your training processes will need to interrupt training, not just rank 0. Otherwise the other processes will crash as in your example.

Personally, I run validation in a totally separate process from the training job.

**Reply by yjiangling** [reply]
> All your training processes will need to interrupt training, not just rank 0. Otherwise the other processes will crash as in your example.
> 
> Personally, I run validation in a totally separate process from the training job.

Oh, I see, thanks a lot.
